### DISCUSSION TITLE
General question related to homogenous and heterogeneous graph.

### DISCUSSION THREAD

**Discussion by amalislam675** [pyg-team/pytorch_geometric#8672] [General]
@rusty1s , I have two separate homogeneous graph and I trained the two separate graph via two separate GNN model. Two separate homogeneous graph means: two graphs belong to same nodes type but from two different sources and every graph have a different set of features associated with them. To improve the performance of the GNN model, I want to fuse these two homogeneous graph along with their features. Can you suggest something related to it, is there a way to fuse the features of two homogenous graph in GNN layer? Secondly, to further improve the performance if attention is added in GNN layer for transductive node classification problem, would it be beneficial. Because, in GAT I found, attention mechanism is beneficial for performance improvement in heterophilic graphs (inductive tasks).

**Comment by rusty1s** [comment]
Does there exist a connection/mapping from nodes between the two separate graphs? If there is no mapping, then it is hard to fuse them. What you can do is, for example, compute graph readouts after each GNN layer and append them to the node features of the other graph.

**Reply by amalislam675** [reply]
@rusty1s , in my case two separate graphs have no mapping. My two graph belong to same node type but from two different sources. If possible, can you refer any link about detail explanation, I am somehow confused in getting this point "compute graph readouts after each GNN layer and append them to the node features of the other graph". Secondly, do you think If I add attention in GNN for homogenous graph will it beneficial. My problem is transductive node classification problem. Previously, I trained separate GNN models on two graphs and then I concatenated the two GNN generated features and then did final prediction. Now for performance improvement I want to do fusion of features with attention if you could suggest something.

**Reply by rusty1s** [reply]
Yes, this is possible. Here is a paper that comes to my mind that describes exactly what you want to do: https://arxiv.org/abs/1904.12787

**Reply by amalislam675** [reply]
@rusty1s, Thankyou. I check this paper.

**Reply by amalislam675** [reply]
@rusty1s , this paper is very helpful, it calculates graph matching between two different graphs. If possible, can you have a look on my GNN model, my GNN model is present in this link is https://docs.google.com/presentation/d/1FAlYUmJuTz9avSXFWyXYM5wFJrwYQYpj/edit?usp=sharing&ouid=117465167918761393377&rtpof=true&sd=true
In this model, I apply GNN model on two separate graphs independently and then simply concatenate and predict my final labels. Indeed, in my case two separate graphs are from same datatype, Protein-protein interaction network graphs but from two different sources. For performance improvement, I want to do feature fusion via attention. Can you tell something, at which place should I embed this Graph Matching Algorithm in to my model? In this paper they apply cross attention mechanism, do you think if two graphs belong to same feature type but from from different sources, does cross attention is beneficial. Which other types of attention mechanisms are beneficial for my task for fusion? If possible, can you provide some useful suggestions as you provided before. Can you have a look on this link, https://openaccess.thecvf.com/content_CVPR_2019/papers/Perez-Rua_MFAS_Multimodal_Fusion_Architecture_Search_CVPR_2019_paper.pdf ? Do you think if I apply fusion based design as is proposed in this paper are beneficial for my task. In this paper multi-model fusion is proposed to fuse image and text features, but in my case, both features belong to same data type, so I am not sure.

**Reply by rusty1s** [reply]
I would apply the cross attention mechanism after every GNN layer at best. If this is not possible due to computational limitations, I would perform it once in the middle of the network such that later GNN layers can still utilize the cross attention.

**Reply by amalislam675** [reply]
@rusty1s , thankyou for your suggestion. You mean I need to apply general cross attention mechanism after every GNN layer in my model in order to further enhance my model performance. Or, I need to compute the similarity between different graphs via graph matching as is done in the paper whose link you have shared above.  I have one another question regarding cross attention mechanism, in general cross attention mechanism we provide query from one source while the key and value from the another source. If I apply the cross attention after first GNN layer then in next GNN layer as an adjacency matrix of the graph I need to provide the adjacency matrix of the source which I have taken as a query for cross attention mechanism application and as a feature I should provide the content vector which I will get after the application of cross attention mechanism and do the same but in reverse for the second source and in this way I should apply the cross attention mechanism for all of the GNN layers. In your opinion,  cross attention mechanism is still beneficial even if two sources belong to same datatype but from different sources. For attention mechanism, in your opinion I can try scaled dot product, dot product, additive, or multiplicative attention mechanism, any one which perform best.

**Reply by rusty1s** [reply]
> You mean I need to apply general cross attention mechanism after every GNN layer in my model in order to further enhance my model performance.

Yes.

> In your opinion, cross attention mechanism is still beneficial even if two sources belong to same datatype but from different sources.

Yes, if I understand it correctly, attention will "learn a mapping" here, and enhances the data sources by information from another data source. It's hard to tell which attention mechanism will perform best. Scaled dot-product is probably the first thing I would try.

**Reply by amalislam675** [reply]
@rusty1s , thanks for your suggestion.
